"Creating a simple plugin" guide (#12022)
ckeditor/ckeditor5: docs/framework/guides/simple-plugin-tutorial/abbreviation-plugin-level-3.md

Start by importing the {@link module:utils/keystrokehandler~KeystrokeHandler `KeystrokeHandler`} and {@link module:utils/focustracker~FocusTracker `FocusTracker`} classes into your form view, and creating their new instances in the `constructor()`.

Now, in the `render()` method, you can add each element of our `childViews` view collection to the `focusTracker`. There, you can also start listening for the keystrokes coming from the rendered view element.

Lastly, let's add a `destroy()` method, and destroy both the focus tracker and the keystroke handler. It will ensure that when the user kills the editor, our helpers 'die' too, preventing any memory leakages.

```js
// abbreviation/abbreviationview.js

import {
	View,
	LabeledFieldView,
	createLabeledInputText,
	ButtonView,
	submitHandler,
} from '@ckeditor/ckeditor5-ui';
import { FocusTracker, KeystrokeHandler } from '@ckeditor/ckeditor5-utils'; // ADDED
import { icons } from '@ckeditor/ckeditor5-core';

export default class FormView extends View {
	constructor( locale ) {
		// ...

		this.focusTracker = new FocusTracker();
		this.keystrokes = new KeystrokeHandler();

		// ...
	}

    render() {
		// ...

> 		this.childViews._items.forEach( v => {

Reviewer: Can we use `view` instead of `v`? I think it would be more meaningful.

</info-box>

In order insert the user's selection in the form field, we need to first grab and concatenate all text from the selected range. If the user selects a couple of paragraphs, a heading, and an image, we need to go through all the nodes, and use only the ones containing text.

Let's create a helper `getRangeText()` function in a separate `/utils.js` file. It will grab all items from a range using its `getItems()` method. Then, it will concatenate all text from {@link module:engine/model/text~Text `text`} and {@link module:engine/model/textproxy~TextProxy `textProxy`} nodes, and skip all others.

```js
// abbreviation/utils.js

// A helper function that retrieves and concatenates all text within the model range.
export default function getRangeText( range ) {
	return Array.from( range.getItems() ).reduce( ( rangeText, node ) => {
		if ( !( node.is( 'text' ) || node.is( 'textProxy' ) ) ) {
			return rangeText;
		}

		return rangeText + node.data;
	}, '' );
}

```

> Now, in `AbbreviationUI` we can adjust the `_showUI()` method to display the selected text in the abbreviation input field. Import `getRangeText` and pass it the first range in the selection (using {@link module:engine/model/documentselection~DocumentSelection#getFirstRange `getFirstRange()`} method).

Reviewer: ```suggestion
Now, in `AbbreviationUI` we can adjust the `_showUI()` method to display the selected text in the abbreviation input field. Import `getRangeText` and pass the first range in the selection (using {@link module:engine/model/documentselection~DocumentSelection#getFirstRange `getFirstRange()`} method) as an argument.
```

		}
	}
	// ...
}
```

Our new functionality should work now, check it out yourself! It doesn't recognize if selected text is an abbreviation already, so if you select "WYSIWYG", the full title doesn't appear in the title input field. We'll change it in a couple of steps.

## Adding a command

Our plugin does what we want it to do, so why complicate things by adding a command? Well, a command not only executes an action, but also automatically reacts when any changes are applied to the model.

<info-box>
A command in CKEditor 5 is a combination of an action and a state. The state of the command refreshes its state whenever anything changes in the model. We highly recommend {@link framework/guides/architecture/core-editor-architecture#commands reading up on commands} before we move on.
</info-box>

When the user makes a selection in the editor, the command will automatically check if there is an abbreviation there. It will also ensure that the command is only enabled only where the "abbreviation" attribute can be set on the current model selection (not on images, for instance).

### Creating a command

Let's start by creating our command, and moving the existing action logic there.

> In the `/abbreviationcommand.js` file import {@link module:core/command~Command the `Command` class}, and create its instance.

Reviewer: ```suggestion
In the `/abbreviationcommand.js` file import the {@link module:core/command~Command `Command`} class, and create its instance.
```

We'll start by simply moving there the action we already created for `submit` in our `_createFormView()` method, passing the title and the abbreviation text into the command's `execute()` method.

```js
// abbreviation/abbreviationcommand.js

import Command from '@ckeditor/ckeditor5-core/src/command';

export default class AbbreviationCommand extends Command {
	execute( { title, abbr } ) {

> 
> 			editor.model.change( writer => {

Reviewer: ```suggestion
			editor.model.change( writer => {
```

				editor.model.insertContent(
					writer.createText( abbr, { abbreviation: title } )
				);
			} );

	}
}
```

> Now, let's add initialize our `AbbreviationCommand`, by adding it to the list of editor's commands in `AbbreviationEditing`. We'll also pass there a name we'll use to call our command.

Reviewer: ```suggestion
Now, let's initialize our `AbbreviationCommand`, by adding it to the list of editor's commands in `AbbreviationEditing`. We'll also pass there a name we'll use to call our command.
```

				title: formView.titleInputView.fieldView.element.value
			};
			editor.execute( 'addAbbreviation', value );

			this._hideUI();
		} );

		// ...
	}

	// ...
}

```
The command should now work, and pressing the `submit` button should have the same effect as it did before. We can now explore some additional functionalities.

### Refreshing the state

Thanks to the command's {@link module:core/command~Command#refresh `refresh()`} method, we can observe the state and the value of our command not just when the user presses the button, but whenever any changes are made in the editor. We'll use this to check if the user's selection has an abbreviation model attribute already.

Before we do that, let's check if the command can be use at all on a given selection. If the user selects an image, the command should be disabled. Let's check if our `abbreviation` attribute is allowed in the schema, using it's {@link module:engine/model/schema~Schema#checkAttributeInSelection `checkAttributeInSelection()`} method.

```js
// abbreviation/abbreviationcommand.js

import Command from '@ckeditor/ckeditor5-core/src/command';

export default class AbbreviationCommand extends Command {

> 		refresh() {

Reviewer: ```suggestion
	refresh() {
```
